src/class/usbtmc/usbtmc_device.h

```diff
@@ -0,0 +1,133 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * This file is part of the TinyUSB stack.
+ */
+
+
+#ifndef CLASS_USBTMC_USBTMC_DEVICE_H_
+#define CLASS_USBTMC_USBTMC_DEVICE_H_
+
+#include "usbtmc.h"
+
+// Enable 488 mode by default
+#if !defined(USBTMC_CFG_ENABLE_488)
+#define USBTMC_CFG_ENABLE_488 (1)
+#endif
+
+// USB spec says that full-speed must be 8,16,32, or 64.
+// However, this driver implementation requires it to be >=32
+#define USBTMCD_MAX_PACKET_SIZE (64u)
+
+/***********************************************
+ *  Functions to be implemeted by the class implementation
+ */
+
+#if (USBTMC_CFG_ENABLE_488)
+extern usbtmc_response_capabilities_488_t const usbtmcd_app_capabilities;
+#else
+extern usbtmc_response_capabilities_t const usbtmcd_app_capabilities;
+#endif
+
+bool usbtmcd_app_msgBulkOut_start(uint8_t rhport, usbtmc_msg_request_dev_dep_out const * msgHeader);
+
+// transfer_complete does not imply that a message is complete.
+bool usbtmcd_app_msg_data(uint8_t rhport, void *data, size_t len, bool transfer_complete);
+
+bool usbtmcd_app_msgBulkIn_request(uint8_t rhport, usbtmc_msg_request_dev_dep_in const * request);
+
+bool usbtmcd_app_msgBulkIn_complete(uint8_t rhport);
+
+// Indicator pulse should be 0.5 to 1.0 seconds long
+TU_ATTR_WEAK bool usbtmcd_app_indicator_pluse(uint8_t rhport, tusb_control_request_t const * msg, uint8_t *tmcResult);
+
+#if (USBTMC_CFG_ENABLE_488)
+uint8_t usbtmcd_app_get_stb(uint8_t rhport, uint8_t *tmcResult);
+TU_ATTR_WEAK bool usbtmcd_app_msg_trigger(uint8_t rhport, usbtmc_msg_generic_t* msg);
+//TU_ATTR_WEAK bool usbtmcd_app_go_to_local(uint8_t rhport);
+#endif
+
+/*******************************************
+ * Called from app
+ *
+ * We keep a reference to the buffer, so it MUST not change until the app is
+ * notified that the transfer is complete.
+ ******************************************/
+
+bool usbtmcd_transmit_dev_msg_data(
+    uint8_t rhport,
+    usbtmc_msg_dev_dep_msg_in_header_t const * hdr,
+    const void *data);
+
+
+/* "callbacks" from USB device core */
+
+bool usbtmcd_open(uint8_t rhport, tusb_desc_interface_t const * itf_desc, uint16_t *p_length);
+void usbtmcd_reset(uint8_t rhport);
+bool usbtmcd_xfer_cb(uint8_t rhport, uint8_t ep_addr, xfer_result_t result, uint32_t xferred_bytes);
+bool usbtmcd_control_request(uint8_t rhport, tusb_control_request_t const * request);
+bool usbtmcd_control_complete(uint8_t rhport, tusb_control_request_t const * request);
+void usbtmcd_init(void);
+
+/************************************************************
+ * USBTMC Descriptor Templates
```

Comment:
all descriptors template must be prefixed with `TUD_` and follows the same naming convention as one in usbd.h
https://github.com/hathach/tinyusb/blob/master/src/device/usbd.h#L164

I forgot the reason why I put them all in one place at usbd.h, look like they are better at their class driver. Anyways, please move it to usbd.h . I will check and possibly move them to their own class later on in a separate PR.

Reply:
Ok, I'll move this right before finalizing the pull request (to minimize merge conflicts).

Reply:
Ok, I'll move this right before finalizing the pull request (to minimize merge conflicts).